src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs

```diff
@@ -247,7 +248,26 @@ public void Insert(int index, T item)
             /// Adds an item to the <see cref="ICollection{T}"/>.
             /// </summary>
             /// <param name="item">The object to add to the <see cref="ICollection{T}"/>.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
How much of the improvement you're showing is due to AggressiveInlining vs due to the changes in the method body?  I'm not convinced this should be AggressiveInlining.

Reply:
# Benchmark

## Notes

| Method | Description |
| -- | -- |
| TweakedAdd | current implementation |
| SplitAdd | this PR |

`_NoInline`-methods are attributed with `[MethodImpl(MethodImplOptions.NoInlining)]`
`__Inline`-methods are attributed with `[MethodImpl(MethodImplOptions.AggressiveInlining)]`
Methos without `_Xxx` are without any attributes.

## Results

``` ini

BenchmarkDotNet=v0.10.11, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.309)
Processor=Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), ProcessorCount=8
Frequency=2742189 Hz, Resolution=364.6722 ns, Timer=TSC
.NET Core SDK=2.1.300-preview3-008384
  [Host]     : .NET Core 2.1.0-preview2-26313-01 (Framework 4.6.26310.01), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-preview2-26313-01 (Framework 4.6.26310.01), 64bit RyuJIT


```
|              Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |
|-------------------- |---------:|----------:|----------:|-------:|---------:|
| TweakedAdd_NoInline | 4.532 us | 0.0900 us | 0.1668 us |   2.00 |     0.10 |
|          TweakedAdd | 2.272 us | 0.0479 us | 0.0813 us |   1.00 |     0.00 |
|   TweakedAdd_Inline | 2.317 us | 0.0464 us | 0.0824 us |   1.02 |     0.05 |
|   SplitAdd_NoInline | 2.699 us | 0.0505 us | 0.0473 us |   1.19 |     0.04 |
|            SplitAdd | 3.034 us | 0.0601 us | 0.0715 us |   1.34 |     0.05 |
|     SplitAdd_Inline | 2.088 us | 0.0416 us | 0.0696 us |   0.92 |     0.04 |

# Discussion

## SplitAdd

The JIT won't inline `SplitAdd` due to `[FAILED: unprofitable inline] Builder:SplitAdd(long):this` which seems strange to me, because the dasm for this method is:
```asm
; Assembly listing for method Builder:SplitAdd(long):this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; fully interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] (  8,  6.50)     ref  ->  rdi         this class-hnd
;  V01 arg1         [V01,T01] (  5,  3.50)    long  ->  rsi        
;  V02 loc0         [V02,T02] (  6,  4   )     int  ->  rax        
;  V03 loc1         [V03,T03] (  5,  4   )     ref  ->  rdx         class-hnd
;# V04 OutArgs      [V04    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]  
;
; Lcl frame size = 0

G_M50053_IG01:

G_M50053_IG02:
       8B4710               mov      eax, dword ptr [rdi+16]
       488B5708             mov      rdx, gword ptr [rdi+8]
       394208               cmp      dword ptr [rdx+8], eax
       760E                 jbe      SHORT G_M50053_IG04
       4863C8               movsxd   rcx, eax
       488974CA10           mov      qword ptr [rdx+8*rcx+16], rsi
       FFC0                 inc      eax
       894710               mov      dword ptr [rdi+16], eax

G_M50053_IG03:
       C3                   ret      

G_M50053_IG04:
       48B8981431A3AC7F0000 mov      rax, 0x7FACA3311498

G_M50053_IG05:
       48FFE0               rex.jmp  rax

; Total bytes of code 39, prolog size 0 for method Builder:SplitAdd(long):this
; ============================================================
```
Really not much code.

So `SplitAdd` isn't inlined, then why `SplitAdd_NoInline` from the benchmark shows different numbers? It's becuase of the different prolog, and the `rex.jmp` (although I have to admit that I don't know what `rex.jmp` is (yeah, I could search for it) and where it comes from):
```asm
; Assembly listing for method Builder:SplitAdd(long):this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] (  8,  6.50)     ref  ->  rdi         this class-hnd
;  V01 arg1         [V01,T01] (  5,  3.50)    long  ->  rsi        
;  V02 loc0         [V02,T02] (  6,  4   )     int  ->  rax        
;  V03 loc1         [V03,T03] (  5,  4   )     ref  ->  rdx         class-hnd
;# V04 OutArgs      [V04    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]  
;
; Lcl frame size = 8

G_M50053_IG01:
       50                   push     rax

G_M50053_IG02:
       8B4710               mov      eax, dword ptr [rdi+16]
       488B5708             mov      rdx, gword ptr [rdi+8]
       394208               cmp      dword ptr [rdx+8], eax
       7612                 jbe      SHORT G_M50053_IG04
       4863C8               movsxd   rcx, eax
       488974CA10           mov      qword ptr [rdx+8*rcx+16], rsi
       FFC0                 inc      eax
       894710               mov      dword ptr [rdi+16], eax

G_M50053_IG03:
       4883C408             add      rsp, 8
       C3                   ret      

G_M50053_IG04:
       E8B4F9FFFF           call     Builder:AddWithResize(long):this
       90                   nop      

G_M50053_IG05:
       4883C408             add      rsp, 8
       C3                   ret      

; Total bytes of code 42, prolog size 1 for method Builder:SplitAdd(long):this
; ============================================================
```

Note: with _AggressiveInling_ the JIT emits a `call` and no `rex.jmp` instruction.

Side note: in https://github.com/dotnet/corefx/pull/28177#discussion_r175297033 it's maybe that there was no _AggressiveInlining_.

## TweakedAdd

JIT will inline this method by default, although the dasm is much greater than the one `SplitAdd` (here the dasm is shown from `TweakAdd_NoInline`):
```asm
; Assembly listing for method Builder:TweakedAdd(long):this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] ( 10, 10   )     ref  ->  rbx         this class-hnd
;  V01 arg1         [V01,T03] (  4,  4   )    long  ->  r14        
;  V02 loc0         [V02,T04] (  4,  4   )     int  ->  r15        
;  V03 tmp0         [V03,T01] (  6, 12   )     ref  ->  rax        
;  V04 tmp1         [V04,T02] (  6, 12   )     int  ->  rdi        
;# V05 OutArgs      [V05    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]  
;
; Lcl frame size = 0

G_M41370_IG01:
       4157                 push     r15
       4156                 push     r14
       53                   push     rbx
       488BDF               mov      rbx, rdi
       4C8BF6               mov      r14, rsi

G_M41370_IG02:
       8B7B10               mov      edi, dword ptr [rbx+16]
       448D7F01             lea      r15d, [rdi+1]
       488BFB               mov      rdi, rbx
       418BF7               mov      esi, r15d
       E893FFFFFF           call     Builder:EnsureCapacity(int):this
       488B4308             mov      rax, gword ptr [rbx+8]
       8B7B10               mov      edi, dword ptr [rbx+16]
       3B7808               cmp      edi, dword ptr [rax+8]
       7312                 jae      SHORT G_M41370_IG04
       4863FF               movsxd   rdi, edi
       4C8974F810           mov      qword ptr [rax+8*rdi+16], r14
       44897B10             mov      dword ptr [rbx+16], r15d

G_M41370_IG03:
       5B                   pop      rbx
       415E                 pop      r14
       415F                 pop      r15
       C3                   ret      

G_M41370_IG04:
       E8B0080F79           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3     

; Total bytes of code 65, prolog size 5 for method Builder:TweakedAdd(long):this
; ============================================================
```

# Conclusion

The "raw implementation" (NoInline compared) of `SplitAdd` is way faster than `TweakedAdd`. Because the JIT won't inline _SplitAdd_ we could
* improve JITs inlining heuristics
* force the method to inline (AggressiveInling)

In [List<T>.Add](https://github.com/dotnet/coreclr/blob/c2ba3e1370e53ca992477c4a19dddf09d9d137bb/src/mscorlib/shared/System/Collections/Generic/List.cs#L225-L240) the similar pattern with fast- and cold-path is used, and there is also _AggressiveInling_. So I don't see any reason why we shouldn't go with that.

Reply:
You mentioned:


| Method | Description |
| -- | -- |
| TweakedAdd | current implementation |
| SplitAdd | this PR |

But if I look at just those two rows of your results table in the same comment, it looks like "this PR" slows down the scenario. Am I reading it right? If so, I don't know why we would take this PR.

Reply:
Theses results are for benchmarks to answert the question about the influence of aggressive inlining.
So the suffixes like `_Inline` have to taken into account when reading the results.

This PR, as implemented, shows the results in the `SplitAdd_Inline` row, and there is an improvement. Not a huge one, but still noticeable faster.